{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "# Path to the JSON file\n",
    "json_file_path = './dataset/BioASQ/test.json'\n",
    "\n",
    "# Initialize lists to store the individual fields\n",
    "test_id_list = []\n",
    "test_question_list = []\n",
    "test_context_list = []\n",
    "test_label_list = []\n",
    "\n",
    "# Open the JSON file and read line by line\n",
    "with open(json_file_path, 'r', encoding='utf-8') as file:\n",
    "    for line in file:\n",
    "        # Parse each line (which is a JSON object) into a Python dictionary\n",
    "        sample = json.loads(line.strip())\n",
    "        \n",
    "        # Extract and store each field in its respective list\n",
    "        test_id_list.append(sample.get('id'))\n",
    "        test_question_list.append(sample.get('sentence1'))\n",
    "        test_context_list.append(sample.get('sentence2'))\n",
    "        test_label_list.append(sample.get('label'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting results from LLMs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import openai\n",
    "import os\n",
    "import pathlib\n",
    "import textwrap\n",
    "import google.generativeai as genai\n",
    "from IPython.display import display\n",
    "from IPython.display import Markdown\n",
    "from google.colab import userdata\n",
    "from vertexai.generative_models import GenerationConfig\n",
    "\n",
    "from transformers import AutoTokenizer, T5ForConditionalGeneration\n",
    "from transformers import LlamaTokenizer, LlamaForCausalLM\n",
    "from transformers import pipeline\n",
    "\n",
    "\n",
    "def get_output(prompt, llm):\n",
    "    if llm == 3.5:\n",
    "        openai.api_key = ''\n",
    "        model = 'gpt-3.5-turbo-1106'\n",
    "        message = openai.ChatCompletion.create(\n",
    "            model=model,\n",
    "            temperature=0,\n",
    "            messages=[\n",
    "                    {\"role\": \"user\", \"content\": prompt}\n",
    "                ]\n",
    "        )\n",
    "        result = message['choices'][0]['message']['content']\n",
    "\n",
    "    elif llm == 4:\n",
    "        openai.api_key = ''\n",
    "        model = 'gpt-4'\n",
    "        message = openai.ChatCompletion.create(\n",
    "            model=model,\n",
    "            temperature=0,\n",
    "            messages=[\n",
    "                    {\"role\": \"user\", \"content\": prompt}\n",
    "                ]\n",
    "        )\n",
    "        result = message['choices'][0]['message']['content']\n",
    "\n",
    "    elif llm == 'instruct':\n",
    "        openai.api_key = ''\n",
    "        model = \"gpt-3.5-turbo-instruct\"\n",
    "        message = openai.Completion.create(\n",
    "            model = model,\n",
    "            prompt = prompt,\n",
    "            temperature = 0    \n",
    "        )\n",
    "        result = message['choices'][0]['text']   \n",
    "    \n",
    "    elif llm == 'gemini-1.0-pro':\n",
    "        GOOGLE_API_KEY = userdata.get('GOOGLE_API_KEY')\n",
    "        genai.configure(api_key=GOOGLE_API_KEY)\n",
    "        model = genai.GenerativeModel('gemini-pro')\n",
    "\n",
    "        safety_settings = [\n",
    "            {\"category\": \"HARM_CATEGORY_SEXUALLY_EXPLICIT\", \"threshold\": \"BLOCK_NONE\"},\n",
    "            {\"category\": \"HARM_CATEGORY_HATE_SPEECH\", \"threshold\": \"BLOCK_NONE\"},\n",
    "            {\"category\": \"HARM_CATEGORY_HARASSMENT\", \"threshold\": \"BLOCK_NONE\"},\n",
    "            {\"category\": \"HARM_CATEGORY_DANGEROUS_CONTENT\", \"threshold\": \"BLOCK_NONE\"}\n",
    "        ]        \n",
    "\n",
    "        response = model.generate_content(prompt,\n",
    "                generation_config=genai.types.GenerationConfig(\n",
    "                candidate_count=1,\n",
    "                temperature=0),\n",
    "                safety_settings=safety_settings)\n",
    "    \n",
    "        result = response.candidates[0].content.parts[0].text\n",
    "    \n",
    "\n",
    "    elif llm == 'flan-ul2':\n",
    "        model = T5ForConditionalGeneration.from_pretrained(\"google/flan-ul2\", device_map=\"auto\", load_in_8bit=True)                                                                 \n",
    "        tokenizer = AutoTokenizer.from_pretrained(\"google/flan-ul2\")\n",
    "        inputs = tokenizer(prompt, return_tensors=\"pt\").input_ids.to(\"cuda\")\n",
    "        outputs = model.generate(inputs, max_length=200)\n",
    "        result = tokenizer.decode(outputs[0])\n",
    "    \n",
    "\n",
    "    elif llm == 'med-alpaca':\n",
    "        pipeline = pipeline(\"text-generation\", model=\"medalpaca/medalpaca-7b\", tokenizer=\"medalpaca/medalpaca-7b\")\n",
    "        result = pipeline(prompt, max_length=200)[0]['generated_text']\n",
    "\n",
    "    elif llm == 'pmc-llama':\n",
    "        tokenizer = LlamaTokenizer.from_pretrained('axiong/PMC_LLaMA_13B')\n",
    "        model = LlamaForCausalLM.from_pretrained('axiong/PMC_LLaMA_13B')\n",
    "        encoded_input = tokenizer(prompt, return_tensors=\"pt\", add_special_tokens=False)\n",
    "        output = model.generate(**encoded_input)\n",
    "        result = tokenizer.decode(output[0])\n",
    "\n",
    "    else:\n",
    "        raise ValueError('Invalid LLM')\n",
    "\n",
    "    print(result)\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_prompt(question, reference, prompt_type='base'):\n",
    "    \n",
    "    if prompt_type == 'guide':\n",
    "        prompt = f'''### Task\n",
    "You are a skilled medical expert. Considering the information from a biomedical study provided in the reference text, is it correct to conclude that '{question}'? Please first respond with 'yes' or 'no', followed by a brief explanation of your reasoning process, ensuring that your explanation aligns with the study's findings.\n",
    "'''\n",
    "    else:\n",
    "        prompt = f'''### Task\n",
    "You are a skilled medical expert. Considering the information from a biomedical study provided in the reference text, is it correct to conclude that '{question}'? Please respond with 'yes' or 'no'.\n",
    "'''\n",
    "\n",
    "\n",
    "    if prompt_type == '1shot':\n",
    "        prompt += f''' \n",
    "### Examples\n",
    "Example Question 1: Is the protein Papilin secreted?\n",
    "Example Reference Text 1: Using expression analysis, we identify three genes that are transcriptionally regulated by HLH-2: the protocadherin cdh-3, and two genes encoding secreted extracellular matrix proteins, mig-6/papilin and him-4/hemicentin.  We found that mig-6 encodes long (MIG-6L) and short (MIG-6S) isoforms of the extracellular matrix protein papilin, each required for distinct aspects of DTC migration. Both MIG-6 isoforms have a predicted N-terminal papilin cassette apilins are homologous, secreted extracellular matrix proteins which share a common order of protein domains.  The TSR superfamily is a diverse family of extracellular matrix and transmembrane proteins, many of which have functions related to regulating matrix organization, cell-cell interactions and cell guidance. This review samples some of the contemporary literature regarding TSR superfamily members (e.g. F-spondin, UNC-5, ADAMTS, papilin, and TRAP) where specific functions are assigned to the TSR domains. Papilins are extracellular matrix proteins  Papilin is an extracellular matrix glycoprotein   Collagen IV, laminin, glutactin, papilin, and other extracellular matrix proteins were made primarily by hemocytes and were secreted into the medium.  A sulfated glycoprotein was isolated from the culture media of Drosophila Kc cells and named papilin.\n",
    "Example Output Text 1: yes\n",
    "'''\n",
    "\n",
    "    elif prompt_type == '3shot':\n",
    "        prompt += f''' \n",
    "### Examples\n",
    "Example Question 1: Is the protein Papilin secreted?\n",
    "Example Reference Text 1: Using expression analysis, we identify three genes that are transcriptionally regulated by HLH-2: the protocadherin cdh-3, and two genes encoding secreted extracellular matrix proteins, mig-6/papilin and him-4/hemicentin.  We found that mig-6 encodes long (MIG-6L) and short (MIG-6S) isoforms of the extracellular matrix protein papilin, each required for distinct aspects of DTC migration. Both MIG-6 isoforms have a predicted N-terminal papilin cassette apilins are homologous, secreted extracellular matrix proteins which share a common order of protein domains.  The TSR superfamily is a diverse family of extracellular matrix and transmembrane proteins, many of which have functions related to regulating matrix organization, cell-cell interactions and cell guidance. This review samples some of the contemporary literature regarding TSR superfamily members (e.g. F-spondin, UNC-5, ADAMTS, papilin, and TRAP) where specific functions are assigned to the TSR domains. Papilins are extracellular matrix proteins  Papilin is an extracellular matrix glycoprotein   Collagen IV, laminin, glutactin, papilin, and other extracellular matrix proteins were made primarily by hemocytes and were secreted into the medium.  A sulfated glycoprotein was isolated from the culture media of Drosophila Kc cells and named papilin.\n",
    "Example Output Text 1: yes\n",
    "\n",
    "Example Question 2: Does metformin interfere thyroxine absorption?\n",
    "Example Reference Text 2: LT4 absorption is unchanged by concomitant metformin ingestion. It has been hypothesized that metformin may suppress serum thyrotropin (TSH) concentrations by enhancing LT4 absorption or by directly affecting the hypothalamic-pituitary axis.\n",
    "Example Output Text 2: no\n",
    "\n",
    "Example Question 3: Do mutations of AKT1 occur in meningiomas?\n",
    "Example Reference Text 3: The recent identification of somatic mutations in components of the SHH-GLI1 and AKT1-MTOR signaling pathways indicates the potential for cross talk of these pathways in the development of meningiomas. A mutation in PIK3CA or AKT1 was found in around 9 % of the cases. AKT1E17K mutations cluster with meningothelial and transitional meningiomas and can be detected by SFRP1 immunohistochemistry. AKT1E17K mutations were exclusively seen in meningiomas and occurred in 65 of 958 of these tumors. A strong preponderance was seen in the variant of meningothelial meningioma WHO grade I of basal and spinal localization. In contrast, AKT1E17K mutations were rare in WHO grade II and absent in WHO grade III meningiomas.  We observed strong up-regulation of SFRP1 expression in all meningiomas with AKT1E17K mutation and in HEK293 cells after transfection with mutant AKT1E17K, but not in meningiomas and HEK293 cells lacking this mutation. SMO and AKT1 mutations occur in non-NF2 meningiomas. Recurrent mutations in SMO and AKT1 are mutually exclusive with NF2 loss in meningioma. Genomic sequencing of meningiomas identifies oncogenic SMO and AKT1 mutations.  A subset of meningiomas lacking NF2 alterations harbored recurrent oncogenic mutations in AKT1 (p.Glu17Lys) and SMO (p.Trp535Leu) and exhibited immunohistochemical evidence of activation of these pathways. Genomic analysis of non-NF2 meningiomas reveals mutations in TRAF7, KLF4, AKT1, and SMO. A subset of meningiomas lacking NF2 alterations harbored recurrent oncogenic mutations in AKT1 (p.Glu17Lys) and SMO (p.Trp535Leu) and exhibited immunohistochemical evidence of activation of these pathways. SMO and AKT1 mutations occur in non-NF2 meningiomas The recent identification of somatic mutations in components of the SHH-GLI1 and AKT1-MTOR signaling pathways indicates the potential for cross talk of these pathways in the development of meningiomas A subset of meningiomas lacking NF2 alterations harbored recurrent oncogenic mutations in AKT1 (p.Glu17Lys) and SMO (p.Trp535Leu) and exhibited immunohistochemical evidence of activation of these pathways Genomic analysis of non-NF2 meningiomas reveals mutations in TRAF7, KLF4, AKT1, and SMO Genomic sequencing of meningiomas identifies oncogenic SMO and AKT1 mutations Recurrent mutations in SMO and AKT1 are mutually exclusive with NF2 loss in meningioma A subset of meningiomas lacking NF2 alterations harbored recurrent oncogenic mutations in AKT1 (p.Glu17Lys) and SMO (p.Trp535Leu) and exhibited immunohistochemical evidence of activation of these pathways. These mutations were present in therapeutically challenging tumors of the skull base and higher grade.  A subset of meningiomas lacking NF2 alterations harbored recurrent oncogenic mutations in AKT1 (p.Glu17Lys) and SMO (p.Trp535Leu) and exhibited immunohistochemical evidence of activation of these pathways.\n",
    "Example Output Text 3: yes\n",
    "'''\n",
    "\n",
    "    elif prompt_type == '5shot':\n",
    "        prompt += f''' \n",
    "### Examples\n",
    "Example Question 1: Is the protein Papilin secreted?\n",
    "Example Reference Text 1: Using expression analysis, we identify three genes that are transcriptionally regulated by HLH-2: the protocadherin cdh-3, and two genes encoding secreted extracellular matrix proteins, mig-6/papilin and him-4/hemicentin.  We found that mig-6 encodes long (MIG-6L) and short (MIG-6S) isoforms of the extracellular matrix protein papilin, each required for distinct aspects of DTC migration. Both MIG-6 isoforms have a predicted N-terminal papilin cassette apilins are homologous, secreted extracellular matrix proteins which share a common order of protein domains.  The TSR superfamily is a diverse family of extracellular matrix and transmembrane proteins, many of which have functions related to regulating matrix organization, cell-cell interactions and cell guidance. This review samples some of the contemporary literature regarding TSR superfamily members (e.g. F-spondin, UNC-5, ADAMTS, papilin, and TRAP) where specific functions are assigned to the TSR domains. Papilins are extracellular matrix proteins  Papilin is an extracellular matrix glycoprotein   Collagen IV, laminin, glutactin, papilin, and other extracellular matrix proteins were made primarily by hemocytes and were secreted into the medium.  A sulfated glycoprotein was isolated from the culture media of Drosophila Kc cells and named papilin.\n",
    "Example Output Text 1: yes\n",
    "\n",
    "Example Question 2: Does metformin interfere thyroxine absorption?\n",
    "Example Reference Text 2: LT4 absorption is unchanged by concomitant metformin ingestion. It has been hypothesized that metformin may suppress serum thyrotropin (TSH) concentrations by enhancing LT4 absorption or by directly affecting the hypothalamic-pituitary axis.\n",
    "Example Output Text 2: no\n",
    "\n",
    "Example Question 3: Do mutations of AKT1 occur in meningiomas?\n",
    "Example Reference Text 3: The recent identification of somatic mutations in components of the SHH-GLI1 and AKT1-MTOR signaling pathways indicates the potential for cross talk of these pathways in the development of meningiomas. A mutation in PIK3CA or AKT1 was found in around 9 % of the cases. AKT1E17K mutations cluster with meningothelial and transitional meningiomas and can be detected by SFRP1 immunohistochemistry. AKT1E17K mutations were exclusively seen in meningiomas and occurred in 65 of 958 of these tumors. A strong preponderance was seen in the variant of meningothelial meningioma WHO grade I of basal and spinal localization. In contrast, AKT1E17K mutations were rare in WHO grade II and absent in WHO grade III meningiomas.  We observed strong up-regulation of SFRP1 expression in all meningiomas with AKT1E17K mutation and in HEK293 cells after transfection with mutant AKT1E17K, but not in meningiomas and HEK293 cells lacking this mutation. SMO and AKT1 mutations occur in non-NF2 meningiomas. Recurrent mutations in SMO and AKT1 are mutually exclusive with NF2 loss in meningioma. Genomic sequencing of meningiomas identifies oncogenic SMO and AKT1 mutations.  A subset of meningiomas lacking NF2 alterations harbored recurrent oncogenic mutations in AKT1 (p.Glu17Lys) and SMO (p.Trp535Leu) and exhibited immunohistochemical evidence of activation of these pathways. Genomic analysis of non-NF2 meningiomas reveals mutations in TRAF7, KLF4, AKT1, and SMO. A subset of meningiomas lacking NF2 alterations harbored recurrent oncogenic mutations in AKT1 (p.Glu17Lys) and SMO (p.Trp535Leu) and exhibited immunohistochemical evidence of activation of these pathways. SMO and AKT1 mutations occur in non-NF2 meningiomas The recent identification of somatic mutations in components of the SHH-GLI1 and AKT1-MTOR signaling pathways indicates the potential for cross talk of these pathways in the development of meningiomas A subset of meningiomas lacking NF2 alterations harbored recurrent oncogenic mutations in AKT1 (p.Glu17Lys) and SMO (p.Trp535Leu) and exhibited immunohistochemical evidence of activation of these pathways Genomic analysis of non-NF2 meningiomas reveals mutations in TRAF7, KLF4, AKT1, and SMO Genomic sequencing of meningiomas identifies oncogenic SMO and AKT1 mutations Recurrent mutations in SMO and AKT1 are mutually exclusive with NF2 loss in meningioma A subset of meningiomas lacking NF2 alterations harbored recurrent oncogenic mutations in AKT1 (p.Glu17Lys) and SMO (p.Trp535Leu) and exhibited immunohistochemical evidence of activation of these pathways. These mutations were present in therapeutically challenging tumors of the skull base and higher grade.  A subset of meningiomas lacking NF2 alterations harbored recurrent oncogenic mutations in AKT1 (p.Glu17Lys) and SMO (p.Trp535Leu) and exhibited immunohistochemical evidence of activation of these pathways.\n",
    "Example Output Text 3: yes\n",
    "\n",
    "Example Question 4: Is enzastaurin effective treatment of glioblastoma?\n",
    "Example Reference Text 4: RESULTS: fourteen randomized clinical trials were identified (7 with bevacizumab, 2 cilengitide, 1 enzastaurin, 1 dasatinib, 1 vandetanib, 1 temsirolimus, 1 cediranib) including 4330 patients. Antiangiogenic drugs showed no improvement in overall survival with a pooled HR of 1.00, a trend for an inferior outcome, in terms of overall survival, was observed in the group of patients receiving antiangiogenic drug alone compared to cytotoxic drug alone (HR=1.24, p=0.056).  Enzastaurin (LY317615) in combination with bevacizumab for recurrent malignant gliomas is well-tolerated, with response and progression-free survival similar to bevacizumab monotherapy. So far, inhibition of angiogenesis by compounds such as bevacizumab, cediranib, enzastaurin or cilengitide as well as alternative dosing schedules of temozolomide did not prolong survival, neither at primary diagnosis nor at recurrent disease. Despite promising phase II clinical trial results and patient benefit in terms of clinical improvement and longer progression-free survival, an overall survival benefit has not been demonstrated in four randomized phase III trials of bevacizumab or cilengitide in newly diagnosed glioblastoma or cediranib or enzastaurin in recurrent glioblastoma. CONCLUSIONS: PFS-6 missed the primary planned outcome of 55%.  OS (median, 74 weeks) and PFS (median, 36 weeks) results from the current trial were comparable to those from a prior phase II study using erlotinib and were significantly better than those from 2 other previous studies that used thalidomide or cis-retinoic acid, all in combination with temozolomide plus RT.  More recently, antiangiogenic agents including enzastaurin, cediranib, bevacizumab, and others that target mainly the VEGF pathway, have been evaluated in this highly angiogenic disease. Among them, only bevacizumab has been associated with clear anti-tumor activity, although the lack of control studies limits the impact of the results to date. Enzastaurin has anti-glioma activity in patients with recurrent high-grade glioma, but does not appear to have enough single-agent activity to be useful as monotherapy. Several signal transduction inhibitors have been examined in preclinical and clinical malignant glioma trials, including antiangiogenic agents (bevacizumab, enzastaurin), and inhibitors of epidermal growth factor receptor tyrosine kinase (gefitinib and erlotinib), mammalian target of rapamycin (temsirolimus, everolimus) and integrin (cilengitide). Although preliminary clinical results of the use of targeted agents have not translated into significantly better survival, more recent phase II trials are exploring the combination of multitargeted drugs with cytotoxic chemotherapy and radiotherapy in order to overcome the resistance of tumors to single-agent targeted therapies. Several drugs have been tested, including epidermal growth factor receptor (EGFR) tyrosine kinase inhibitors (gefitinib and erlotinib), mammalian target of rapamycin (mTOR) inhibitors (temsirolimus and everolimus), and vascular endothelial growth factor receptor (VEGFR), protein kinase C-beta, and other angiogenesis pathways inhibitors (vatalanib, bevacizumab, and enzastaurin). Although preliminary efficacy results of most trials in recurrent disease have fallen short on expectations, substantial advances have been achieved by associated translational research.  So far, inhibition of angiogenesis by compounds such as bevacizumab, cediranib, enzastaurin or cilengitide as well as alternative dosing schedules of temozolomide did not prolong survival, neither at primary diagnosis nor at recurrent disease.\n",
    "Example Output Text 4: no\n",
    "\n",
    "Example Question 5: Does echinacea increase anaphylaxis risk?\n",
    "Example Reference Text 5: Chicoric acid (dicaffeoyl-tartaric acid), is a natural phenolic compound found in a number of plants, such as chicory (Cichorium intybus) and Echinacea (Echinacea purpurea), which possesses antioxidant, anti-inflammatory, antiviral, and analgesic activities. Although these biological effects of chicoric acid have been investigated, there are no reports of its antiallergic-related anti-inflammatory effects in human mast cells (HMC)-1 or anaphylactic activity in a mouse model. BACKGROUND: Fifty percent of Australians use complementary and alternative medicines (other than vitamins) in any 12-month period, of which echinacea-containing products are increasingly popular. Recent reports have highlighted the risk of allergic reactions to complementary medicines in atopic patients. Two patients suffered anaphylaxis and a third had an acute asthma attack 10 minutes after their first ever dose of echinacea. Fifty-one Australian adverse drug reports implicating echinacea were also reviewed. There were 26 cases suggestive of possible immunoglobulin E-mediated hypersensitivity (4 anaphylaxis, 12 acute asthma, 10 urticaria/angioedema).  Echinacea-associated anaphylaxis. A woman with atopy experienced anaphylaxis after taking, among other dietary supplements, a commercial extract of echinacea. Risk of anaphylaxis in complementary and alternative medicine. Several culprits identified including Andrographis paniculata, Echinacea species, bee products, Ginkgo biloba and Ginseng are discussed here.SUMMARY: Knowing the factors that increase the risk of anaphylaxis allows reactions to be recognized, reported and further investigated.\n",
    "Example Output Text 5: yes\n",
    "'''\n",
    "\n",
    "    # Add input text to be annotated\n",
    "    prompt += f'''\n",
    "### Reference Text: {reference}\n",
    "### Output Text:\n",
    "'''\n",
    "\n",
    "    #print(prompt)\n",
    "    return prompt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run(llm, prompt_type, ids, questions, contexts):\n",
    "    output_path = f'./output/BioASQ/{llm}/'\n",
    "    if not os.path.exists(output_path):\n",
    "        os.makedirs(output_path)\n",
    "\n",
    "    # Check if the file exists and delete it before starting to append new outputs\n",
    "    output_file_path = os.path.join(output_path, f'test_{prompt_type}.json')\n",
    "    if os.path.exists(output_file_path):\n",
    "        os.remove(output_file_path)\n",
    "    \n",
    "    for id, question, context in zip(ids, questions, contexts): \n",
    "        success = False\n",
    "        while not success:\n",
    "            try:\n",
    "                # Assume get_output is a function that generates the output\n",
    "                prompt = create_prompt(question, context, prompt_type)\n",
    "                output = get_output(prompt, llm)\n",
    "\n",
    "                # todo: model may not generate proper answer??\n",
    "                cleaned_output = ' '.join(output.split())\n",
    "\n",
    "                # Create a json string\n",
    "                output_dict = {\"id\": id, \"answer\": cleaned_output}\n",
    "                json_string = json.dumps(output_dict)\n",
    "                \n",
    "                # Open the file in append mode ('a') to add each new output\n",
    "                with open(output_file_path, 'a', encoding='utf-8') as f_write:\n",
    "                    f_write.write(json_string + '\\n')\n",
    "                    success = True\n",
    "                    \n",
    "            except Exception as e:\n",
    "                print(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "llm = 4\n",
    "prompt_type = 'base'\n",
    "run(llm, prompt_type, test_id_list, test_question_list, test_context_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evaluation (accuracy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "# load results from specified dir\n",
    "llm = 'gemini'\n",
    "prompt_type = '3shot'\n",
    "output_file_path = f'./output/BioASQ/{llm}/test_{prompt_type}.json'\n",
    "\n",
    "test_id2_list = []\n",
    "test_answer_list = []\n",
    "\n",
    "# Open the JSON file and read line by line\n",
    "with open(output_file_path, 'r', encoding='utf-8') as file:\n",
    "    for line in file:\n",
    "        sample = json.loads(line.strip())\n",
    "\n",
    "        # Extract sample id\n",
    "        sample_id = sample.get('id')\n",
    "        test_id2_list.append(sample_id)\n",
    "\n",
    "        # Use regular expression to find 'Yes' or 'No' at the beginning of the text\n",
    "        sample_answer = sample.get('answer')\n",
    "        extracted_answer_match = re.search(r\"^\\s*(Yes|No)\", sample_answer, re.IGNORECASE)\n",
    "        extracted_answer = extracted_answer_match.group(0) if extracted_answer_match else \"Answer not found!\"\n",
    "        test_answer_list.append(extracted_answer)\n",
    "\n",
    "assert test_id_list == test_id2_list\n",
    "\n",
    "print(test_answer_list)\n",
    "print(test_label_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import accuracy_score, roc_auc_score\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "print(\"Evaluating: \", llm, prompt_type)\n",
    "\n",
    "# Convert labels and predictions to lowercase\n",
    "labels = [label.lower() for label in test_label_list]\n",
    "predictions = [prediction.lower() for prediction in test_answer_list]\n",
    "\n",
    "# Convert labels to binary values (0 and 1)\n",
    "le = LabelEncoder()\n",
    "labels_binary = le.fit_transform(labels)\n",
    "predictions_binary = le.transform(predictions)\n",
    "\n",
    "# Calculate accuracy\n",
    "accuracy = accuracy_score(labels_binary, predictions_binary)\n",
    "print(\"Accuracy:\", accuracy)\n",
    "\n",
    "# Calculate AUC score (may not get correct result with few samples)\n",
    "try: \n",
    "    auc_score = roc_auc_score(labels_binary, predictions_binary)\n",
    "    print(\"AUC Score:\", auc_score)\n",
    "except:\n",
    "    pass"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pytorch",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
